test/test_vmap.py

```diff
@@ -1150,5 +1160,100 @@ def test_no_random_op_support(self):
                                         'vmap: We do not yet support calling random operations'):
                 vmap(op)(*args)
 
+def construct_v(output, batch_size):
+    return torch.randn(batch_size, *output.shape,
+                       dtype=output.dtype, device=output.device)
+
+def as_tuple(x):
+    if istuple(x):
+        return x
+    elif isinstance(x, list):
+        return tuple(x)
+    else:
+        return x,
+
+def differentiable(args):
+    return tuple(arg for arg in as_tuple(args)
+                 if isinstance(arg, torch.Tensor) and arg.requires_grad)
+
+class TestVmapBatchedGradient(TestCase):
+    def _vmap_test(self, *args, **kwargs):
+        return _vmap_test(self, *args, **kwargs)
+
+    # Tests batched gradient computation of outputs = op(*args, **kwargs)
+    # by comparing it to a sequential map+stack fallback.
+    #
+    # output_process_fn: a function that maps the outputs to the part
+    #       that should be differentiated.
+    # batch_size: the batch dim size for the batched grad
+    def _batched_grad_test(self, op, args, kwargs, output_process_fn=lambda x: x, batch_size=3):
+        outputs = op(*args, **kwargs)
+        outputs = differentiable(output_process_fn(outputs))
+        batched_vectors = tuple(construct_v(out, batch_size) for out in outputs)
+
+        def vector_jacobian_product(*vectors):
+            return torch.autograd.grad(outputs, differentiable(args), vectors,
+                                       retain_graph=True)
+        self._vmap_test(vector_jacobian_product, batched_vectors,
+                        check_propagates_grad=False)
+
+    # Tests batched second grad computation of outputs = op(*args, **kwargs).
+    # by comparing it to a sequential map+stack fallback.
+    #
+    # output_process_fn: a function that maps the outputs to the part
+    #       that should be differentiated.
+    # batch_size: the batch dim size for the batched grad
+    #
+    # NB: we only test computing batched gradients are in the second gradient
+    # computation. One specific use case that does this is computing the hessian
+    # matrix of a scalar-valued function; this is useful in Bayesian Logistic
+    # Regression.
+    # It might be useful to have a test that computes batched first gradients and
+    # then uses those to compute batched second gradients in the future.
+    def _batched_grad_grad_test(self, op, args, kwargs, output_process_fn=lambda x: x, batch_size=3):
```

Comment:
One question is: I could make this (and _batched_grad_grad_test) a part of gradcheck, but I'm not sure if I should.

The nice thing about making it a part of gradcheck is that we can be sure that "batched gradient" works for literally all operators in PyTorch. The downside of making it a part of gradcheck right now is modifying a lot of callsites of gradcheck to not check this (because _batched_grad_grad_test fails for a lot of operators).

I think in the long run we should have gradcheck test for batched gradient, but it might take a while to get there

Reply:
We have a sort of similar situation with complex numbers. My current thinking is that it's fine to not make it part of gradcheck in the beginning when it's only a handful of ops that support batched gradient. Then as you start adding support fo more operators, at some point you will get to the point where you are going for "completion"; at that point, it makes sense to flip the switch, and go from whitelisting ops which do work, to blacklisting ops that don't work.